### PR TITLE
Support for incremental protocols through ETL

### PR DESCRIPTION
**Pull request by ankurjuneja**

### CODE REVIEW

File: nirc_ehr/resources/queries/dbo/q_protocol.sql [hunk 1]
```diff
@@ -34,6 +34,44 @@ SELECT 'PROTOCOL-' || PROTOCOL_ID AS "objectid",
...
        REVIEW_COMPLETION_DATE AS "reviewCompletionDate",
-       HONOR_RESET_YN AS "isHonorReset"
-FROM PROTOCOL
+       HONOR_RESET_YN AS "isHonorReset",
+       COALESCE(MAX(CAST(adt.CHANGE_DATETIME AS TIMESTAMP)), to_date('01/01/1970' ,'MM/DD/YYYY')) AS modified
+FROM PROTOCOL p
+LEFT JOIN AUDIT_TRAIL adt ON p.PROTOCOL_ID = substring(PRIMARY_KEY_VALUES, length('PROTOCOL_ID = '))
+AND adt.TABLE_NAME = 'PROTOCOL'
 WHERE PROTOCOL_NUMBER IS NOT NULL
+GROUP BY PROTOCOL_ID,
```
Comment (labkey-jeckels) [inline nirc_ehr/resources/queries/dbo/q_protocol.sql]: Long GROUP BYs are bad for query perf and maintenance.

Here, you are only joining to AUDIT_TRAIL to get the MAX(adt.CHANGE_DATETIME). There's no need to do the left join (expanding the row count based on the AUDIT_TRAIL rows) and then collapse it back (which forces the DB to sort on ALL of the fields in the GROUP BY).

Instead, make this part of the LEFT JOIN. Don't simply join AUDIT_TRAIL, join to a subquery that does a GROUP BY on substring(PRIMARY_KEY_VALUES, length('PROTOCOL_ID = ')) and SELECTs the substring and the MAX(adt.CHANGE_DATETIME)

File: nirc_ehr/resources/queries/dbo/q_protocol_amendments.sql [hunk 2]
```diff
@@ -3,6 +3,14 @@ SELECT PROTOCOL_ID AS "protocolId",
...
-FROM PROTOCOL
-WHERE PROTOCOL_TYPE_ID IN (2,3,4)
\ No newline at end of file
+       APPROVAL_DATE AS "approved",
+       COALESCE(MAX(CAST(adt.CHANGE_DATETIME AS TIMESTAMP)), to_date('01/01/1970' ,'MM/DD/YYYY')) AS modified
+FROM PROTOCOL p
+LEFT JOIN AUDIT_TRAIL adt ON p.PROTOCOL_ID = substring(PRIMARY_KEY_VALUES, length('PROTOCOL_ID = '))
+AND adt.TABLE_NAME = 'PROTOCOL'
+WHERE PROTOCOL_TYPE_ID IN (2,3,4)
+GROUP BY PROTOCOL_ID,
```
Comment (labkey-jeckels) [inline nirc_ehr/resources/queries/dbo/q_protocol_amendments.sql]: While the GROUP BY list is much shorter here, I recommend using the same approach I describe above.

**Review by labkey-jeckels — CHANGES_REQUESTED** [review]

File: nirc_ehr/resources/queries/dbo/q_protocol.sql (hunk 1 above)
Comment (ankurjuneja) [reply nirc_ehr/resources/queries/dbo/q_protocol.sql]: done

File: nirc_ehr/resources/queries/dbo/q_protocol_amendments.sql (hunk 2 above)
Comment (ankurjuneja) [reply nirc_ehr/resources/queries/dbo/q_protocol_amendments.sql]: done

**Review by labkey-jeckels — APPROVED** [review]

**Pull request merged**